### PR TITLE
EJB App configuration finalization [ECR-1936] [ECR-1935] [ECR-1770]

### PR DESCRIPTION
**Pull request by vitvakatu**
## Overview

- ECR-1936 `java.library.path` parameter has been removed from the configuration, because the combination of `LD_LIBRARY_PATH` and `RUNPATH` is enough to load native library at runtime.
- ECR-1935 system classpath is now located at runtime (in `./lib/java` directory relatively to EJB App location)
- ECR-1770 all configuration parameters now fit requirements.

---
See: https://jira.bf.local/browse/ECR-1936
See: https://jira.bf.local/browse/ECR-1935
See: https://jira.bf.local/browse/ECR-1770

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] The [continuous integration build](https://www.travis-ci.org/exonum/exonum-java-binding) passes

### CODE REVIEW

**Comment by vitvakatu** [comment]
Changes in `start_cryptocurrency_node` are temporary - you cannot run node without some preparation at the moment, so don't take it into account.

File: exonum-java-binding-core/rust/src/runtime/config.rs [hunk 1]
```diff
@@ -22,23 +22,19 @@ pub struct JvmConfig {
     pub system_class_path: String,
     /// Java service classpath.
     pub service_class_path: String,
-    /// Path to java-bindings shared library.
-    ///
-    /// Should be path to exonum-java-binding-core/rust/target/{debug, release}
-    pub lib_path: String,
```
Comment (skletsun) [inline exonum-java-binding-core/rust/src/runtime/config.rs]: Also integration tests seems not adopted for removed `lib_path` and `ServiceConfig.port`

Comment (vitvakatu) [reply exonum-java-binding-core/rust/src/runtime/config.rs]: Oh yep, indeed :upside_down_face:

File: exonum-java-binding-core/rust/src/runtime/cmd.rs [hunk 2]
```diff
@@ -122,29 +137,48 @@ impl CommandExtension for Finalize {
...
+        let file = entry.unwrap();
+        if file.file_type().unwrap().is_file() {
+            jars.push(file.path());
+        } else {
+            continue;
+        }
+    }
+
+    let jars: Vec<&str> = jars.iter().map(|p| p.to_str().unwrap()).collect();
+    join_paths(&jars)
```
Comment (ilammy) [inline exonum-java-binding-core/rust/src/runtime/cmd.rs]: Isn't it possible to use `join_paths` without an intermediate Vec? Like,

```rust
join_paths(jars.iter())
```

Comment (vitvakatu) [reply exonum-java-binding-core/rust/src/runtime/cmd.rs]: Due to `join_paths` signature, not.

Comment (ilammy) [reply exonum-java-binding-core/rust/src/runtime/cmd.rs]: Oh, I thought that was `std::env::join_paths`.

Comment (vitvakatu) [reply exonum-java-binding-core/rust/src/runtime/cmd.rs]: :thinking: Why we're not using `env::join_paths`?

Comment (ilammy) [reply exonum-java-binding-core/rust/src/runtime/cmd.rs]: vOv

`utils::join_paths` seems to have been added in #331, but I believe that's exactly what `std::env::join_paths` does.

Comment (vitvakatu) [reply exonum-java-binding-core/rust/src/runtime/cmd.rs]: Got it, `utils::join_paths` is used for different purpose and can't be replaced with `std::env::join_paths`

File: exonum-java-binding-core/rust/src/runtime/cmd.rs [hunk 3]
```diff
@@ -1,97 +1,53 @@
...
-            system_class_path,
-            service_class_path,
-            lib_path,
-            log_config_path,
-        };
+        let public_config = PublicConfig { module_name };
 
-        let mut services_secret_configs = context
-            .get(keys::SERVICES_SECRET_CONFIGS)
+        let mut services_public_configs = context
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/src/runtime/cmd.rs]: I'd consider splitting into methods or at least sections with comments clarifying the steps.

I assume it is 
// Create a public config of this service and node? — BTW, why each node has to specify module name? Shan't it be global config?

// Put this service config at this node — But how at this stage can we have several of them?

File: exonum-java-binding-core/rust/src/runtime/cmd.rs [hunk 4]
```diff
@@ -100,15 +56,73 @@ impl CommandExtension for GenerateNodeConfig {
...
+            true,
+            "Java service classpath.",
+            None,
+            "ejb-service-classpath",
+            false,
+        )]
+    }
+
+    fn execute(&self, mut context: Context) -> Result<Context, failure::Error> {
+        let service_class_path = context.arg(EJB_SERVICE_CLASSPATH)?;
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/src/runtime/cmd.rs]: I'd split or add clarifying comments.

File: exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh [hunk 5]
```diff
@@ -62,16 +62,16 @@ ejb-app generate-template --validators-count=1 testnet/common.toml
...
- --ejb-libpath $EJB_LIBPATH \
- --ejb-log-config-path $EJB_LOG_CONFIG_PATH \
+ --ejb-module-name 'com.exonum.binding.cryptocurrency.ServiceModule' \
  --peer-address 127.0.0.1:5400
 
 header "FINALIZE"
 ejb-app finalize testnet/sec.toml testnet/node.toml \
- --ejb-module-name 'com.exonum.binding.cryptocurrency.ServiceModule' \
- --ejb-port 6000 \
+ --ejb-service-classpath $EJB_CLASSPATH \
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: according to configuration requirements

[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: That does not answer the question _why_ — what is the reasoning behind this decision?

[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: It's private parameter, so there is no reason to put it before finalize step

**Comment by vitvakatu** [comment]
Now module name is set at TEMPLATE step and correctly shared among every node.

File: exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh [hunk 6]
```diff
@@ -37,10 +39,8 @@ echo "PROJ_ROOT=${EJB_ROOT}"
 
 header "PREPARE PATHS"
 
-CORE_TXT="exonum-java-binding-core/target/ejb-core-classpath.txt"
 CRYPTOCURRENCY_TXT="exonum-java-binding-cryptocurrency-demo/target/cryptocurrency-classpath.txt"
-EJB_CLASSPATH="$(cat ${EJB_ROOT}/${CORE_TXT}):$(cat ${EJB_ROOT}/${CRYPTOCURRENCY_TXT})"
-EJB_CLASSPATH="${EJB_CLASSPATH}:${EJB_ROOT}/exonum-java-binding-core/target/classes"
+EJB_CLASSPATH="$(cat ${EJB_ROOT}/${CRYPTOCURRENCY_TXT})"
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: Doesn't really matter at the moment, script still can't be used to launch App without some manual action

File: exonum-java-binding-core/rust/src/runtime/cmd.rs [hunk 7]
```diff
@@ -164,12 +165,16 @@ impl CommandExtension for Run {
 fn get_system_classpath() -> String {
     let mut jars = Vec::new();
     let jars_directory = {
+        // Get current path to EJB App.
         let mut current_directory =
-            env::current_dir().expect("Could not get current working directory");
+            env::current_exe().expect("Could not get current working directory");
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/src/runtime/cmd.rs]: Could not get the executable directory

File: exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: It has to be done — I'd fire an issue if not here.

[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: Agreed, let's do it as a part of packaging-related activities.

File: exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding-core/rust/ejb-app/start_cryptocurrency_node.sh]: Yeah, classpath — exactly, I seem to have commented on the module name which is the same on all nodes (public and not node-specific).

**Comment by vitvakatu** [comment]
Mm.... Wait a bit, I'll update `ejb-app-configuration` branch first

**Comment by vitvakatu** [comment]
Sorry, was forced to do a force push. Looks okay, let's wait for CI

**Pull request merged**